### ISSUE TITLE
Ability to set defaultType on pack files as well as override type on a specific file

### ISSUE THREAD

**Issue by jrabek** [PhaserEditor2D/PhaserEditor2D-v3#237]
In order to more fully leverage the built in loading system of Phaser, it would be convenient if Phaser Editor allowed you to specify the `defaultType` for a json pack file (https://github.com/photonstorm/phaser/blob/master/src/loader/LoaderPlugin.js#L650) as well as override the `type` property for a particular file.

The reason is that if you want to load assets via a pack file that are not built-in Phaser  assets, you have to load the pack file via a new pack loader plugin.  In the new pack loader plugin you have to process the pack json and create instances of the particular file type that you want.  This is due to the fact that the built in pack loader tries to look up the file's type after the pack file is  processed and won't be able to find the new type.

To clarify with an example, in this case we want to load Rive animations ([rive.app](https://rive.app/)), but when they are added to a pack file they get set with a generic type of `bin`.  This means that while the file gets loaded, it won't be processed until after the preload phase since there is no generic `bin` type that would do the processing during the load phase (https://github.com/photonstorm/phaser/blob/master/src/loader/LoaderPlugin.js#L665).

By allowing the defaultType and type to be overridden, we could either place all rive animations in a single pack file and specify the `defaultType` to `rive` or place the rive animations in the main pack file and just set the `type` property to be `rive` instead of `bin`.  And we would just need to create a simple `RiveFile` class that extends `Phaser.Loader.File`.

**Comment by PhaserEditor2D** [comment]
Well, if a file type is not supported by the Asset Pack editor, then you should load it by code. Writing the loading code in the `Preloader.ts` file.

But if yet you need to load it in the asset pack, then the way is to create a plugin for extending the asset pack.
I can help on creating this plugin.

**Comment by jrabek** [comment]
So I don't want to load each rive file individually in the final game.  I would prefer to just include the rive files in I need in the asset pack and just pass one asset pack to the preloader. We have multiple games and so the preloader is shared between them via submodule (along with other shared components.

There are two ways we can do this:

1) (hacky way) Create a separate pack file of just the Rive animations that is processed by a specialized loader plugin (i.e. RivePackLoader) that will load the json pack file and assume every file is  a Rive file.  The problem is that you have to repeat the json parsing code (which isn't a lot but it isn't the way the Phaser loader system was meant to be used for pack files.  The downside here is that we have to have a separate pack file and loader and can only put Rive animations in the pack file.

2) (Phaser way) Set the `type` property on the file to be `rive` and not just the generic `bin` that gets set when adding the files in the loader and the phaser loader will automatically use the custom loader method to load just that file.

```
function riveLoaderCallback(this: Phaser.Loader.LoaderPlugin, font: string, url: string): Phaser.Loader.LoaderPlugin {

  const fileConfig: RivFileConfig = {
    type: "rive",
    key: url,
    url,
    extension: "riv",
  };
  this.addFile(new RivFile(this, fileConfig));

  return this;
}

export class RiveFileLoaderPlugin extends Phaser.Plugins.BasePlugin {
  constructor(pluginManager: Phaser.Plugins.PluginManager) {
    super(pluginManager);

    pluginManager.registerFileType("rive", riveLoaderCallback);
  }
}

declare module "phaser" {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace Loader {
    interface LoaderPlugin {
      rive(key: string, url?: string): this;
    }
  }
}
```

**Comment by PhaserEditor2D** [comment]
I prefer option 2, in addition to creating a plugin for the editor so it will support the `rive` file type. Installing plugins to the editor is very simple these days. It is just about installing an NPM module.

Right now the Asset Pack Editor is not open to plugins, but I think it is not hard to open it. So I can work on this and even make a plugin template for future custom types.

I can work on this for the next release.

Please, can you attach here some rive files for testing?

**Comment by jrabek** [comment]
Oh wow, having a proper plugin for the editor would be amazing.  If nothing else just to get the preview of the animation inside the editor would be huge! 🔥 

Here is an example (https://codesandbox.io/s/rive-plain-js-sandbox-1ddrc?file=/src/index.js) of how the high level rive api is used along with a sample animation. 

Note that for the game at run time (as opposed to in the editor), Rive's high level API isn't ideal if many instances of a game object are loaded from the same rive file since the rive file is parsed each time (https://github.com/rive-app/rive-wasm/blob/744f0c1571ddc86ecaa48a2fa4c6b006396dbf1e/js/src/rive.ts#L1117).   That's why I am creating a custom Rive object that is able to instance rive objects.   I already reached out to the Rive team and they said they are improving their API around this and asked me to open an issue (https://github.com/rive-app/rive-wasm/issues/262).

For full support in the editor I imagine you would have to do deep integration and make it rive aware.

However, if time is pressing, I think just being able to modify the file type in the inspector would at least simplify the loading.  For placing the rive objects in the scene I had originally just planned on using white rectangles (just to set the position and bounds) with custom code for the rive object.
<img width="1160" alt="image" src="https://user-images.githubusercontent.com/1304735/191035148-4790764c-6d7e-4787-a530-04ce06d80c65.png">

**Comment by PhaserEditor2D** [comment]
Good. Yes, we can work first on the asset pack editor support, and later on the preview plugin.

The `codesandbox.io` demo is not opening well. Please, can you attach the files here? Just the files to be loaded, not a running game.

**Comment by jrabek** [comment]
Here is the code:
```
import "./styles.css";

const rive = require("@rive-app/canvas");

new rive.Rive({
  // Hosted .riv asset, or a local one. Uncomment line 8 to try with
  // a local basketball.riv asset, or add your own!
  src: "https://cdn.rive.app/animations/off_road_car_v7.riv",
  // src: "basketball.riv",
  canvas: document.getElementById("canvas"),
  autoplay: true
});
```

**Comment by jrabek** [comment]
Github won't let me attach riv files directly to the issue it seems.

**Comment by PhaserEditor2D** [comment]
Oh, sorry, is opening the `codesandbox.io` now, thanks!

**Comment by jrabek** [comment]
@PhaserEditor2D this is great!  Will check this out tomorrow!

**Comment by jrabek** [comment]
@PhaserEditor2D is it possible to install this as part of the editor instead of part of the project?

I ask because in our setup we use a monorepo with npm workspaces for each game that includes the server, client, and shared code (e.g. network message definitions).

When installing the Rive plugin it ends up in the top level `node_modules` since that is how npm workspaces work.  The problem is that Phaser Editor is only opened in the client directory since it is only for editing the client and that way it doesn't have to load all the other project files which makes editing faster in general.

**Comment by jrabek** [comment]
@PhaserEditor2D nevermind! I forgot about the `.phasereditor2d` directory in the home folder.  Seems to be working now.

**Comment by PhaserEditor2D** [comment]
Yes, also you can configure it via a flag `-plugins`, in the `phasereditor2d.config.json`: https://help-v3.phasereditor2d.com/misc/plugins.html#project-s-plugins
